### PR TITLE
[#17] Controller for input data

### LINKED ISSUE #17: Create a Router to receive input data

### PR DESCRIPTION
**Pull request by marsofandrew**
close #17 
Add Receiver for POST /api/rest/conference

### CODE REVIEW

File: api/src/main/java/org/confeti/controllers/Controllers.java [hunk 1]
```diff
@@ -3,4 +3,7 @@
 public final class Controllers {
     public final static String REST_API_PATH = "/api/rest";
 
+    private Controllers(){
+      throw new UnsupportedOperationException();
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/Controllers.java]: I prefer to just leave a comment `not instantiable`. Joshua Bloch from Effective Java [offers](https://stackoverflow.com/questions/25658330/why-java-util-objects-private-constructor-throws-assertionerror) to throw `AssertionError` and include a comment.

File: api/src/main/java/org/confeti/controllers/receiver/ReceiverController.java [hunk 2]
```diff
@@ -1,25 +1,46 @@
...
+    private final ReportService reportService;
+    private final ConferenceService conferenceService;
+
     @PostMapping(consumes = APPLICATION_JSON_VALUE, produces = APPLICATION_JSON_VALUE)
     @ResponseBody
-    public Mono<Status> handlePost(@RequestBody Conference conference) {
-        System.out.println(conference);
-        throw new ResponseStatusException(HttpStatus.NOT_IMPLEMENTED);
+    public Mono<ResponseEntity<Status>> handlePost(@RequestBody final InputData inputData) {
+        return Mono.justOrEmpty(inputData.getConference())
```
Comment (lapter57) [inline api/src/main/java/org/confeti/controllers/receiver/ReceiverController.java:34]: You can do this pipeline in one line:
```
Mono.justOrEmpty(inputData.getConference())
         .switchIfEmpty(Mono.error(BAD_REQUEST_EXCEPTION))
         .map(InpuData::getReports)
         .flatMapMany(Flux::fromIterable)
         ....
         .onErrorReturn(ResponseEntity.badRequest().body(Status.fail()))
```

File: api/src/main/java/org/confeti/dataobjects/Status.java [hunk 3]
```diff
@@ -1,5 +1,10 @@
 package org.confeti.dataobjects;
 
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+@Data
+@AllArgsConstructor
 public class Status {
```
Comment (lapter57) [inline api/src/main/java/org/confeti/dataobjects/Status.java]: You can create enum without outer class [see](https://www.baeldung.com/jackson-serialize-enums)

File: api/src/test/java/org/confeti/controllers/ControllersTests.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package org.confeti.controllers;
+
+public final class ControllersTests {
+    public static final String STATUS = "{\"status\":\"OK\"}";
+
+
+    private ControllersTests() {
+        throw new UnsupportedOperationException();
```
Comment (lapter57) [inline api/src/test/java/org/confeti/controllers/ControllersTests.java]: I prefer to just leave a comment `not instantiable`. Joshua Bloch from Effective Java [offers](https://stackoverflow.com/questions/25658330/why-java-util-objects-private-constructor-throws-assertionerror) to throw `AssertionError` and include a comment.

File: api/src/main/java/org/confeti/service/dto/Speaker.java [hunk 5]
```diff
@@ -28,7 +31,7 @@
 public class Speaker implements Serializable {
 
     private static final long serialVersionUID = 1L;
-
+    @JsonIgnore
```
Comment (lapter57) [inline api/src/main/java/org/confeti/service/dto/Speaker.java]: ` @JsonIgnore` is used to ignore the logical property used in serialization and deserialization. But we need this field in serialization. And I think we should add the test that this field is ignored during the deserialization.

File: api/src/main/java/org/confeti/service/dto/Report.java [hunk 6]
```diff
@@ -30,6 +33,7 @@
 
     private static final long serialVersionUID = 1L;
 
+    @JsonIgnore
```
Comment (lapter57) [inline api/src/main/java/org/confeti/service/dto/Report.java]: @JsonIgnore is used to ignore the logical property used in serialization and deserialization. But we need this field in serialization. And I think we should add the test that this field is ignored during the deserialization.

File: api/src/main/java/org/confeti/controllers/receiver/ReceiverController.java (hunk 2 above)
Comment (marsofandrew) [reply api/src/main/java/org/confeti/controllers/receiver/ReceiverController.java:34]: Really? It doesn't work

Comment (lapter57) [reply api/src/main/java/org/confeti/controllers/receiver/ReceiverController.java:34]: Sorry, I mean:
```
        return Mono.justOrEmpty(inputData.getConference())
                .switchIfEmpty(Mono.error(new Exception("bad request")))
                .doOnNext(conferenceService::upsert)
                .then(Mono.just(inputData))
                .map(data -> Objects.requireNonNullElse(data.getReports(), new ArrayList<Report>()))
                .flatMapMany(Flux::fromIterable)
                .map(report -> report.setConferences(Set.of(inputData.getConference())))
                .flatMap(reportService::upsert)
                .then()
                .thenReturn(ResponseEntity.ok(Status.ok()))
                .onErrorReturn(ResponseEntity.badRequest().body(Status.fail()));
```

but if we add default value of reports in InputData, we can replace `.then(Mono.just(inputData)).map(data -> Objects.requireNonNullElse(data.getReports(), new ArrayList<Report>()))` on `.then(Mono.just(inputData.getReports()))`

**Review by lapter57 — APPROVED** [review]

**Pull request merged**